modules/balancer-sdk/balancer-sor.service.ts

```diff
@@ -46,67 +52,81 @@ export class BalancerSorService {
+                // Chances are a 304 means that we missed a pool draining event, and the pool data is stale.
+                // We force an update on any pools inside of the swapInfo
+                if (error.error?.error?.message.includes('BAL#304')) {
+                    const poolIds = _.uniq(swapInfo.swaps.map((swap) => swap.poolId));
 
-        const returnAmount = formatFixed(
-            swapInfo.returnAmount,
-            this.getTokenDecimals(swapType === 'EXACT_IN' ? tokenOut : tokenIn, tokens),
-        );
+                    Sentry.captureException(
+                        `Received a BAL#304 during getSwaps, forcing an on-chain refresh for: ${poolIds.join(',')}`,
+                    );
+
+                    const blockNumber = await this.provider.getBlockNumber();
+
+                    poolService.updateOnChainDataForPools(poolIds, blockNumber).catch();
+                }
+            }
+
+            throw new Error(error);
+        }
 
         const pools = await poolService.getGqlPools({
             where: { idIn: swapInfo.routes.map((route) => route.hops.map((hop) => hop.poolId)).flat() },
         });
 
-        const tokenInAmount = swapType === 'EXACT_IN' ? swapAmount : returnAmount;
-        const tokenOutAmount = swapType === 'EXACT_IN' ? returnAmount : swapAmount;
+        const tokenInAmount = BigNumber.from(deltas[swapInfo.tokenAddresses.indexOf(tokenIn)]);
+        const tokenOutAmount = BigNumber.from(deltas[swapInfo.tokenAddresses.indexOf(tokenOut)]).abs();
+
+        const swapAmountQuery = swapType === 'EXACT_OUT' ? tokenOutAmount : tokenInAmount;
+        const returnAmount = swapType === 'EXACT_IN' ? tokenOutAmount : tokenInAmount;
 
-        const effectivePrice = oldBnum(tokenInAmount).div(tokenOutAmount);
-        const effectivePriceReversed = oldBnum(tokenOutAmount).div(tokenInAmount);
+        const returnAmountFixed = formatFixed(
+            returnAmount,
+            this.getTokenDecimals(swapType === 'EXACT_IN' ? tokenOut : tokenIn, tokens),
+        );
+
+        const swapAmountQueryFixed = formatFixed(
+            swapAmountQuery,
+            this.getTokenDecimals(swapType === 'EXACT_OUT' ? tokenOut : tokenIn, tokens),
+        );
+
+        const tokenInAmountFixed = formatFixed(tokenInAmount, this.getTokenDecimals(tokenIn, tokens));
+        const tokenOutAmountFixed = formatFixed(tokenOutAmount, this.getTokenDecimals(tokenOut, tokens));
+
+        const effectivePrice = oldBnum(tokenInAmountFixed).div(tokenOutAmountFixed);
+        const effectivePriceReversed = oldBnum(tokenOutAmountFixed).div(tokenInAmountFixed);
         const priceImpact = effectivePrice.div(swapInfo.marketSp).minus(1);
 
+        for (const route of swapInfo.routes) {
+            route.tokenInAmount = oldBnum(tokenInAmountFixed).multipliedBy(route.share).toString();
+            route.tokenOutAmount = oldBnum(tokenOutAmountFixed).multipliedBy(route.share).toString();
```

Comment:
on both of these, you're going to end up with more decimal places then you started with, breaking any calls to parseUnits later on. This needs adjusting.

Reply:
You need to add a call to `dp` before toString()

oldBnum(tokenInAmountFixed).multipliedBy(route.share).dp(numDecimalsForThisToken).toString()